### DISCUSSION TITLE
D-connecting paths

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#15] [Q&A]
@phwuil Do you know if there is a way to output all d-connecting paths from **X** to **Y** relative **Z** given a PDAG in pyagrum once`cSeparation(X, Y, Z)` returns true?

**Comment by phwuil** [comment]
It is an interesting but unusual question. So there is no direct way to find them. The question is to find all the paths between X and Y in the moralized ancestral graph which are not using nodes in Z. Unfortunately, `PDAG.moralizedAncestralGraph()` is not wrapped (yet) in pyAgrum. So you have to do it by yourself...

**Reply by kenneth-lee-ch** [reply]
How come I can see that this function `PDAG.moralizedAncestralGraph()` is available under PDAG documentation?

**Reply by phwuil** [reply]
Creating a wrapper is partially automatic with swig. So the function has been wrapped but we have to write a piece of code in order to correctly transform the parameters. If you try to use it, you will see a C++ error about the argument that has to be a `gum::NodeSet` (or something like that). We have to transform a python sequence of int into this NodeSet. I will try to do that for this function a.s.a.p and you will find it in the pyAgrum-nightly package (you are doomed to stay on a development version :-)

**Reply by phwuil** [reply]
![image](https://github.com/agrumery/aGrUM/assets/5980636/16c5beb8-a4a2-4991-a0d8-a4b8c1d3c34a)

This should be done in pyAgrum-nightly.

**Comment by phwuil** [comment]
By the way, for your initial question, just a first answer ... may not be correct, but it is the idea :

```python
def DFS(g,src,target):
    """
    Parameters
    -----------------
    g: undirected graph
    src,target : nodeId
    
    Returns
    -----------
    List[List[int]] : a list of paths (as list of nodes) from src to target in g.
    """
    def recDFS(g,src,target,current,mark,res):
        if src==target:
            res.append(current)
        elif src not in mark:
            mark[src]=1
            for n in g.neighbours(src):
                recDFS(g,n,target,current+[n],mark,res)
            del(mark[src])
    
    res=[]
    current=[src]
    mark={}
    recDFS(g,src,target,current,mark,res)
    return res
 ```
 



For instance :
 
 
![image](https://github.com/agrumery/aGrUM/assets/5980636/b84f58fe-f692-4c2a-a379-fd43b18065ee)

**Reply by kenneth-lee-ch** [reply]
Very good! That's why I like this package so much! How did you take the conditioning set `Z` into account in your DFS?

**Reply by phwuil** [reply]
it is not in the DFS,  it is before : you call DFS on a mutilated MAG

```python
def csep(x,y,z=[]):
    cs=pdag.cSeparation(x,y,z)
    res=f"pdag.cSeparation({x},{y},{z if len(z)>0 else ''})={cs}"
    mag=pdag.moralizedAncestralGraph([x]+[y]+z)
    # mutilation of mag
    for n in z:
        mag.eraseNode(n)
    if not cs:
        explain=DFS(mag,x,y)
    else:
        explain="-"
    gnb.sideBySide(res,mag,"<br/>".join([str(l) for l in explain]),captions=["csperation","mutilated MAG","paths"],valign="middle")
    
csep(1,9)
csep(1,9,[2])
```

**Reply by kenneth-lee-ch** [reply]
If you don't mind, would you mind explaining the intuition of using moralizedAncestralGraph?

**Reply by phwuil** [reply]
This is the generalization of the d-separation(in a DAG)/separation (in an undirected graph)/C-separation (in a PDAG) :

X and Y are separated by Z if all the paths from X to Y in MAG(X+Y+Z) contains node(s) of Z. (if X and Y are not connected in the MAG mutilated by Z).

You can find many references to this on internet (https://arxiv.org/abs/1302.3606v1)

**Reply by kenneth-lee-ch** [reply]
@phwuil  In your first example, `[1,3,4,6,7,8,9]` and `[1,3,4,6,9]` don't seem to the d-connecting paths for x=1, y= 9, z= 2 in the original graph. Also, I think 2 should be included in the path.

**Reply by phwuil** [reply]
These are paths in the moralized ancestral graph ... Particularly, they do not include  v-structures that are part of Z (or one of their descendants). 

For instance, the path `1-3-4-6-9` in the moralized ncestral graph represent the path `1->2<-3-4<-6<-8-9` in the CPDAG (`2` is in Z, and then  `9` is connected to `8` by moralization, as ancestors of `2`)

**Reply by kenneth-lee-ch** [reply]
Thanks for the reply! I am not familiar enough with the moralized ancestral graph idea. How can I possibly tweak your code to have the return paths be the d-connecting paths that can be directly read off from the original graph e.g. `1->2<-3-4<-6<-8-9`?

**Reply by phwuil** [reply]
it is complicated. You have to differentiate in the path the real 'edges' and the edges due to moralization (fill-ins). And when the path use such a fill-in, you have to figure out in the PDAG why it exists and then re-build the "true" path in the PDAG.

**Comment by kenneth-lee-ch** [comment]
It would be great if this can be part of pyagrum functions in the future!

**Reply by phwuil** [reply]
The function that gives all the path ? 
Theoretically, this list can be very large.  It is a risk to have such a function that could consume so much time and memory in a library.

**Reply by kenneth-lee-ch** [reply]
When should I expect to see the next release that has `moralizedAncestralGraph` available :) ?

**Reply by phwuil** [reply]
:-) not so far in june.

pyAgrum-nightly has been correctly updated ?

**Reply by kenneth-lee-ch** [reply]
Is it the latest release `1.8.1.9.dev202305261685000343`?

**Reply by phwuil** [reply]
yes

**Reply by phwuil** [reply]
1.8.2 is out ! :-)
